python/src/iceberg/table/schema.py

```diff
@@ -0,0 +1,286 @@
+            field_id = name_index_lower[name_or_id.lower()]
+        return self.id_index[field_id]
+
+    def find_type(self, name_or_id: str | int, case_sensitive: bool = True) -> IcebergType:
+        if isinstance(name_or_id, int):
+            return self.id_index[name_or_id].type
+        if case_sensitive:
+            field_id = self.name_index[name_or_id]
+        else:
+            name_index_lower = {name.lower(): field_id for name, field_id in self.name_index.items()}
+            field_id = name_index_lower[name_or_id.lower()]
+        return self.id_index[field_id].type
+
+    def find_column_name(self, column_id: int) -> str:
+        matched_column = [name for name, field_id in self.name_index.items() if field_id == column_id]
+        if not matched_column:
+            raise ValueError(f"Cannot find column name: {column_id}")
+        return matched_column[0]
+
+    def select(self, names: List[str], case_sensitive: bool = True) -> "Schema":
+        if case_sensitive:
+            return self._case_sensitive_select(schema=self, names=names)
+        return self._case_insensitive_select(schema=self, names=names)
+
+    @classmethod
+    def _case_sensitive_select(cls, schema: "Schema", names: List[str]):
+        # TODO: Add a PruneColumns schema visitor and use it here
+        raise NotImplementedError()
+
+    @classmethod
+    def _case_insensitive_select(cls, schema: "Schema", names: List[str]):
+        # TODO: Add a PruneColumns schema visitor and use it here
+        raise NotImplementedError()
+
+
+class SchemaVisitor(Generic[T], ABC):
+    def before_field(self, field: NestedField) -> None:
+        pass
+
+    def after_field(self, field: NestedField) -> None:
+        pass
+
+    def before_list_element(self, element: NestedField) -> None:
+        self.before_field(element)
+
+    def after_list_element(self, element: NestedField) -> None:
+        self.after_field(element)
+
+    def before_map_key(self, key: NestedField) -> None:
+        self.before_field(key)
+
+    def after_map_key(self, key: NestedField) -> None:
+        self.after_field(key)
+
+    def before_map_value(self, value: NestedField) -> None:
+        self.before_field(value)
+
+    def after_map_value(self, value: NestedField) -> None:
+        self.after_field(value)
+
+    @abstractmethod
+    def schema(self, schema: Schema, struct_result: T) -> T:
+        ...
+
+    @abstractmethod
+    def struct(self, struct: StructType, field_results: List[T]) -> T:
+        ...
+
+    @abstractmethod
+    def field(self, field: NestedField, field_result: T) -> T:
+        ...
+
+    @abstractmethod
+    def list(self, list_type: ListType, element_result: T) -> T:
+        ...
+
+    @abstractmethod
+    def map(self, map_type: MapType, key_result: T, value_result: T) -> T:
+        ...
+
+    @abstractmethod
+    def primitive(self, primitive: PrimitiveType) -> T:
+        ...
+
+
+def visit(obj, visitor: SchemaVisitor[T]) -> T:
```

Comment:
Minor: I think we could refactor the visit method to use `@singledispatch`:

```python
@singledispatch
def visit(obj, visitor: SchemaVisitor[T]) -> T:
    raise NotImplementedError("Cannot visit unsupported type: %s" % obj)

@visit.register(Schema)
def visit(schema: Schema, visitor: SchemaVisitor[T]) -> T:
    return visitor.schema(schema, visit(schema.as_struct(), visitor))

...
```

Reply:
Great idea, done! Looks much better